app/models/product_for_cart.rb

```diff
@@ -0,0 +1,84 @@
+class ProductForCart
+
+  include TextHelpers::Translation
+
+  attr_accessor :error_message
+  attr_accessor :error_path
+
+  def initialize(product)
+    @product = product
+  end
+
+  def purchasable_by?(acting_user, session_user)
+    raise NUCore::PermissionDenied unless product.is_accessible_to_user?(session_user)
+    return false if acting_user.blank?
+
+    if !product.available_for_purchase?
```

Comment:
This method is super complex. Is there any way we can break it up?

I don't know if something like this would be any better?

```ruby
[
available_for_purchase,
associatated_price_policies,
...
].none?

def available_for_purchase
  if product.available_for_purchase?
    true
  else
    @error_message = text(".not_available", i18n_params)
    false
  end
end
```

Reply:
Yeah, that’s fair. I felt the same way, but after spending so much time refactoring this code I needed a stopping point at which I could implement the new checks.

Do you think it merits a few more cycles to extract these checks into separate methods? I like that direction, for sure.